_posts/2015-02-11-abstracting-the-frontend-in-a-shared-nothing-architecture-driven-by-functional-teams.md

```diff
@@ -0,0 +1,167 @@
+
+Luckily, I had done most of that long time before, when I initiated the project "tagmanager integration" for our old webshop. So I  started writing down the important KPIs and metrics based on that historical data and the trackingpixels from our tracking solution and our recommendation engine. I summed it all up in a nice table in our wiki, structured the table based on verticals and added a nice description for all keys. Then I wrote the integration stories for the vertical teams so they could integrate the right metatags and declarations in the markup. Awesome.
+
+Well, at least it felt awesome until I looked at the actual implementation done by the teams. The problem was that I haven't been very explicit in declaring the types (and especially the formatting) for the individual metrics. I simply defined something like *on the product detail page we need a product object with the following fields: name, price, category, ...*. Even though I also defined some examples, this still led to fairly diverse implementations. Especially the price formatting was a problem, because there were multiple interpretations about how a price should be expressed (e.g.`"29,90€"` vs. `29.9` which are both valid representations of the same price value).
+
+## Typization to the rescue
+Since Javascript (and JSON, too) is a very loosely typed language, we needed to define abstract types (or interfaces if you're using [Typescript](http://typescriptlang.org)) that explicitly define how values have to be supplied to the DAL. That resulted in various fancy types, e.g. *DALPageData*, *DALUserData* or *DALProductData* to just name a few. Let's look at the type definition for the *DALPageData* type:
+
+{% highlight javascript %}
+// @interface IDALPageData
+DALPageData = {
+  
+  /**
+   * Display name of the page, e.g. 'Homepage' (according to list in table XXX)
+   */
+  "name" : {
+    "type" : "String",
+    "mandatory" : true
+  },
+  
+  /**
+   * Unique page type for this page, e.g. 'checkout-basket'.
+   */
+  "type" : {
+    "type" : "String",
+    "mandatory" : true
+  }
+  
+}
+{% endhighlight %}
+
+As you can see we provide a JSON structure defining the type and some other, optional fields (e.g. *mandatory*, *deprecated*, *apiVersion* and some more). This way we ensure backwards compatibility throughout the API because the teams can safely adapt to new API versions without introducing breaking changes.
+
+The curious reader might ask himself: *"Hey, what's the purpose of defining these as Javascript obejcts?"* Well, within the metatags the data ist supplied as plain object literals. But in our test infrastructure (based on Selenium) we can now run over the entire page, take the above type definitions and compare them to the actual implementation. If there are mismatches, e.g. because a vertical didn't implement a type correctly, we raise an error and get a notification. Now, ain't that sweet?
```

Comment:
Leave the *himself* out. No need to gender the text if it doesn't add value to the text.